### PR TITLE
View Votes

### PR DESCRIPTION
**Pull request by Sjmarf**
Lets admins view the votes of a post/comment. Moderators can also view the votes on versions later than 0.19.3 (no such version exists yet, but [support for this was merged in January](https://github.com/LemmyNet/lemmy/pull/4392)). Admins can ban users from a context menu on each row. 

Unfortunately, there's no way of seeing whether a user is already banned from the community in this menu. [I've opened an issue for this on Lemmy's side](https://github.com/LemmyNet/lemmy/issues/4561).

I also added a divider between the regular and moderator actions in the `UserView` ellipsis menu. To do this, I had to stop using `.secondaryAction`, which necessitated a small refactor.

### CODE REVIEW

File: Mlem/Models/Content/Post/PostModel+MenuFunctions.swift [hunk 1]
```diff
@@ -148,6 +148,14 @@ extension PostModel {
                     await self.notifier.add(.success("\(self.post.locked ? "L" : "Unl")ocked post"))
                 }
             })
+            // TODO: 0.19 deprecation
+            if siteInformation.isAdmin || ((siteInformation.version ?? .infinity) > .init("0.19.3")) {
```
Comment (EricBAndrews) [inline Mlem/Models/Content/Post/PostModel+MenuFunctions.swift:152]: 0.18 deprecation?

If `siteInformation` is null, I think assuming the actions are unavailable would be safer. If we want to keep this for debugging on local instances, we can put a little `#debug` block into default to `.infinity` in debug mode

File: Mlem/Views/Shared/Comments/Comment Item Logic.swift [hunk 2]
```diff
@@ -298,63 +299,75 @@ extension CommentItem {
...
-                    modToolTracker.banUserFromCommunity(
-                        .init(from: hierarchicalComment.commentView.creator),
-                        from: .init(from: hierarchicalComment.commentView.community),
-                        bannedFromCommunity: creatorBannedFromCommunity,
-                        shouldBan: !creatorBannedFromCommunity,
-                        postTracker: nil,
-                        commentTracker: commentTracker
-                    )
-                })
+            // TODO: 0.19 deprecation
```
Comment (EricBAndrews) [inline Mlem/Views/Shared/Comments/Comment Item Logic.swift:305]: 0.18?

File: Mlem/Views/Shared/Moderation/Tools/BanUserView+Logic.swift [hunk 3]
```diff
@@ -79,13 +82,27 @@ extension BanUserView {
...
+                        } else if banFromInstance {
                             comment.commentView.creator.banned = shouldBan
                         } else {
                             comment.commentView.creatorBannedFromCommunity = shouldBan
                         }
                     }
                 }
+                
+                if let votesTracker {
+                    if let index = votesTracker.votes.firstIndex(where: {$0.id == user.userId}) {
```
Comment (EricBAndrews) [inline Mlem/Views/Shared/Moderation/Tools/BanUserView+Logic.swift:96]:
```suggestion
                if let index = votesTracker?.votes.firstIndex(where: {$0.id == user.userId}) {
```

File: Mlem/Views/Shared/Moderation/Tools/BanUserView+Logic.swift [hunk 4]
```diff
@@ -79,13 +82,27 @@ extension BanUserView {
...
+                        if contentRemovalType == .purge {
+                            votesTracker.votes.remove(at: index)
+                        } else if banFromInstance {
+                            votesTracker.votes[index].user.banned = shouldBan
+                        } else {
+                            votesTracker.votes[index].creatorBannedFromCommunity = shouldBan
+                        }
+                    }
```
Comment (EricBAndrews) [inline Mlem/Views/Shared/Moderation/Tools/BanUserView+Logic.swift:104]: Goes with above
```suggestion
                    
```

File: Mlem/Views/Shared/Moderation/Tools/VotesListView.swift [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                        .contextMenu {
+                            ForEach(menuFunctions(for: item)) { item in
+                                MenuButton(menuFunction: item, menuFunctionPopup: $menuFunctionPopup)
+                            }
+                        }
+                        .onAppear {
+                            if item.id == votesTracker.votes.last?.id {
+                                votesTracker.loadNextPage()
+                            }
+                        }
```
Comment (EricBAndrews) [inline Mlem/Views/Shared/Moderation/Tools/VotesListView.swift:69]: *nit*

This pyramid's getting pretty deep--extract to @ViewBuilder function?

**Review by EricBAndrews — CHANGES_REQUESTED** [review]
Functionality looks good, just a couple minor nits 👍

File: Mlem/Models/Content/Post/PostModel+MenuFunctions.swift (hunk 1 above)
Comment (Sjmarf) [reply Mlem/Models/Content/Post/PostModel+MenuFunctions.swift:152]: > 0.18 deprecation?

The feature is only available on instances running versions released after 0.19.3, but is unavailable on 0.19.0, 0.19.1 and 0.19.2. Therefore, we can remove the check when 0.19.2 is deprecated, hence the `// TODO: 0.19 deprecation` comment.

**Comment by Sjmarf** [comment]
I've added support for the `creatorBannedFromCommunity` field on `APIVoteView`, which [they added](https://github.com/LemmyNet/lemmy/pull/4568).

**Review by EricBAndrews — APPROVED** [review]
LGTM! 🚀

**Pull request merged**